crates/miden-testing/src/kernel_tests/tx/test_fpi.rs

```diff
@@ -1073,6 +1073,168 @@ async fn test_nested_fpi_cyclic_invocation() -> anyhow::Result<()> {
+    // unique procedure which calls the second foreign account via FPI and then returns
+    let first_foreign_account_code_source = r#"
+        use.miden::tx
+        use.std::sys
+
+        export.first_account_foreign_proc
+            # pad the stack for the `execute_foreign_procedure` execution
+            padw padw padw push.0.0.0
+            # => [pad(15)]
+
+            # get the hash of the `second_account_foreign_proc` from the advice stack
+            adv_push.4
+
+            # push the ID of the second foreign account from the advice stack
+            adv_push.2
+            # => [foreign_account_id_prefix, foreign_account_id_suffix, FOREIGN_PROC_ROOT, pad(15)]
+
+            # call the second foreign account
+            exec.tx::execute_foreign_procedure
+            # => [result_from_second]
+
+            # keep the result and drop any padding if present
+            exec.sys::truncate_stack
+        end
+    "#;
+
+    let first_foreign_account_component = AccountComponent::compile(
+        NamedSource::new("first_foreign_account", first_foreign_account_code_source),
+        TransactionKernel::with_kernel_library(source_manager.clone()),
+        vec![],
+    )?
+    .with_supports_all_types();
+
+    let first_foreign_account = AccountBuilder::new(ChaCha20Rng::from_os_rng().random())
+        .with_auth_component(Auth::IncrNonce)
+        .with_component(first_foreign_account_component.clone())
+        .build_existing()?;
+
+    // ------ NATIVE ACCOUNT ---------------------------------------------------------------
+    let native_account = AccountBuilder::new(ChaCha20Rng::from_os_rng().random())
+        .with_auth_component(Auth::IncrNonce)
+        .with_component(MockAccountComponent::with_empty_slots())
+        .storage_mode(AccountStorageMode::Public)
+        .build_existing()?;
+
+    let mut mock_chain = MockChainBuilder::with_accounts([
+        native_account.clone(),
+        first_foreign_account.clone(),
+        second_foreign_account.clone(),
+    ])?
+    .build()?;
+    mock_chain.prove_next_block()?;
+
+    let foreign_account_inputs = vec![
+        mock_chain
+            .get_foreign_account_inputs(first_foreign_account.id())
+            .expect("failed to get foreign account inputs"),
+        mock_chain
+            .get_foreign_account_inputs(second_foreign_account.id())
+            .expect("failed to get foreign account inputs"),
+    ];
+
+    // push the hash of the second foreign procedure and the second account ID onto the advice stack
+    // so the first foreign account can call it dynamically
+    let mut advice_inputs = AdviceInputs::default();
+    advice_inputs
+        .stack
+        .extend(*second_foreign_account.code().procedures()[1].mast_root());
+    advice_inputs.stack.extend([
+        second_foreign_account.id().suffix(),
+        second_foreign_account.id().prefix().as_felt(),
+    ]);
```

Comment:
We should be able to remove these lines once refactored to use `procref` as suggested in another comment.